core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1023,24 +1023,38 @@ class GroupCoordinator(val brokerId: Int,
 
   def tryCompleteHeartbeat(group: GroupMetadata, memberId: String, isPending: Boolean, heartbeatDeadline: Long, forceComplete: () => Boolean) = {
     group.inLock {
-      if (isPending) {
+      // The group has been unloaded and invalid, we should complete the heartbeat.
+      if (group.is(Dead)) {
+        forceComplete()
+      } else if (isPending) {
         // complete the heartbeat if the member has joined the group
         if (group.has(memberId)) {
           forceComplete()
         } else false
-      }
-      else {
-        val member = group.get(memberId)
-        if (member.shouldKeepAlive(heartbeatDeadline) || member.isLeaving) {
+      } else {
+        if (shouldCompleteNonPendingHeartbeat(group, memberId, heartbeatDeadline)) {
           forceComplete()
         } else false
       }
     }
   }
 
+  def shouldCompleteNonPendingHeartbeat(group: GroupMetadata, memberId: String, heartbeatDeadline: Long): Boolean = {
+    if (group.has(memberId)) {
+      val member = group.get(memberId)
+      member.shouldKeepAlive(heartbeatDeadline) || member.isLeaving
+    } else {
+      error(s"Member id $memberId was not found in ${group.groupId} during heartbeat expiration, " +
+        s"this could be potentially related to KAFKA-8896")
```

Comment:
Do we need to mention the JIRA in the error message?

Reply:
Yeah, I agree we should drop the JIRA from the message. Also, I'm not sure I see why this should have a different log level than the message in `onExpireHeartbeat`. Either way, the delayed heartbeat is completing after the member was already removed. Maybe we can make this INFO level also?